daemon/containerd/image_push.go

```diff
@@ -2,15 +2,277 @@ package containerd
+) (map[digest.Digest]distributionSource, error) {
+	mountableBlobs := map[digest.Digest]distributionSource{}
+	source, err := getDigestSource(ctx, store, target.Digest)
+
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return nil, err
+		}
+		logrus.WithField("target", target).Debug("distribution source label not found")
+		return mountableBlobs, nil
+	}
+
+	handler := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		_, err := store.Info(ctx, desc.Digest)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, errdefs.System(errors.Wrapf(err, "failed to get metadata of content %s", desc.Digest.String()))
+			}
+
+			if canBeMounted(desc, targetRef, i.registryService.IsInsecureRegistry, source) {
+				mountableBlobs[desc.Digest] = source
+				jobs.Add(desc)
+			}
+			return nil, nil
+		}
+
+		return containerdimages.Children(ctx, store, desc)
+	}
+
+	err = containerdimages.Dispatch(ctx, containerdimages.HandlerFunc(handler), limiter, target)
+	if err != nil {
+		return nil, err
+	}
+
+	return mountableBlobs, nil
+}
+
+func getDigestSource(ctx context.Context, store content.Manager, digest digest.Digest) (distributionSource, error) {
+	source := distributionSource{}
+
+	info, err := store.Info(ctx, digest)
+	if err != nil {
+		if cerrdefs.IsNotFound(err) {
+			return source, errdefs.NotFound(err)
+		}
+		return source, errdefs.System(err)
+	}
+
+	s := extractDistributionSource(info.Labels)
+	if s == nil {
+		return source, errdefs.NotFound(fmt.Errorf("label %q is not attached to %s", labelDistributionSource, digest.String()))
+	}
+
+	return *s, nil
+}
+
+const labelDistributionSource = "containerd.io/distribution.source."
+
+func extractDistributionSource(labels map[string]string) *distributionSource {
+	var source distributionSource
+
+	// Check if this blob has a distributionSource label
+	// if yes, read it as source
+	for k, v := range labels {
+		registry := strings.TrimPrefix(k, labelDistributionSource)
+		if registry != k {
+			ref, err := reference.ParseNamed(registry + "/" + v)
+			if err != nil {
+				continue
+			}
+
+			source.registryRef = ref
+			return &source
+		}
+	}
+
+	return nil
+}
+
+type distributionSource struct {
+	registryRef reference.Named
+}
+
+func (source distributionSource) GetReference(dgst digest.Digest) (reference.Named, error) {
+	return reference.WithDigest(source.registryRef, dgst)
+}
+
+// shouldDownload returns if the given descriptor can be cross-repo mounted
+// when pushing it to a remote reference ref.
+func canBeMounted(desc ocispec.Descriptor, targetRef reference.Named, isInsecureFunc func(string) bool, source distributionSource) bool {
+	mediaType := desc.MediaType
+
+	if containerdimages.IsManifestType(mediaType) {
+		return false
+	}
+	if containerdimages.IsIndexType(mediaType) {
+		return false
+	}
+	if containerdimages.IsConfigType(mediaType) {
+		return false
+	}
```

Comment:
This one is a little over-zealous -- manifest objects _are_ special (indexes and image manifests), but config objects are just regular blobs (like layers), so if they already exist, they're 100% fair game for cross-blob mounts :eyes:

Reply:
Ah, indeed. I was sure that configs are uploaded as manifests too, but this was probably some mistake when developing the initial fork version. Thanks!